hydromt/components/config.py

```diff
@@ -0,0 +1,215 @@
+        else:
+            return self._data
+
+    def _initialize(self, skip_read=False) -> None:
+        """Initialize the kernel configs."""
+        if self._data is None:
+            self._data = dict()
+            if not skip_read:
+                self.read()
+
+    @hydromt_step
+    def write(
+        self,
+        path: Optional[str] = None,
+    ) -> None:
+        """Write kernel config at <root>/{path}."""
+        self._root._assert_write_mode()
+        if self.data:
+            if path is not None:
+                p = join(self._root.path, path)
+            else:
+                p = join(self._root.path, self._config_fn)
+
+            write_path = join(self._root.path, p)
+            self._model.logger.info(f"Writing kernel config to {write_path}.")
+            makedirs(dirname(write_path), exist_ok=True)
+
+            write_data = make_config_paths_relative(self.data, self._root.path)
+            ext = splitext(p)[-1]
+            if ext in [".yml", ".yaml"]:
+                write_yaml(write_path, write_data)
+            elif ext == ".toml":
+                write_toml(write_path, write_data)
+            else:
+                raise ValueError(f"Unknown file extention: {ext}")
+
+        else:
+            self._model.logger.debug("No kernel config found, skip writing.")
+
+    @hydromt_step
+    def read(self, path: Optional[str] = None) -> None:
+        """Read kernel config at <root>/{path}."""
+        self._initialize(skip_read=True)
+        # if path is abs, join will just return path
+        if path is not None:
+            p = path
+        else:
+            p = self._config_fn
+        read_path = join(self._root.path, p)
+        if isfile(read_path):
+            self._model.logger.info(f"Reading kernel config file from {read_path}.")
+        else:
+            self._model.logger.warning(
+                f"No default kernel config was found at {read_path}. It wil be initialized as empty dictionary"
+            )
+            return
+
+        ext = splitext(p)[-1]
+        if ext in [".yml", ".yaml"]:
+            self._data = read_yaml(read_path)
+        else:
+            raise ValueError(f"Unknown file extention: {ext}")
+
+    @hydromt_step
+    def set(self, data: Dict[str, Any]):
+        """Set the config dictionary at key(s) with values.
+
+        Parameters
+        ----------
+        data: Dict[str,Any]
+            A dictionary with the values to be set. keys can be dotted like in
+            :py:meth:`~hydromt.components.config.ConfigComponent.set_value`
+
+        Examples
+        --------
+        >> self.set({'a': 1, 'b': {'c': {'d': 2}}})
+        >> self.data
+            {'a': 1, 'b': {'c': {'d': 2}}}
+        >> self.set({'a.d.f.g': 1, 'b': {'c': {'d': 2}}})
+        >> self.data
+            {'a': {'d':{'f':{'g': 1}}}, 'b': {'c': {'d': 2}}}
+        """
+        for k, v in data.items():
+            self.update(k, v)
```

Comment:
Sounds like a good place for pydantic.v1.utils.deep_update? Or the recursive version that you wrote yourself in another PR.

Reply:
There is this new function `rgetattr` that I built that does exactly that expansion. That might help, right?

Reply:
it would for the `get` that's true, but for the update the dictionaries need to be created which it doesn't do I believe